src/kernel/bitcoinkernel.h

```diff
@@ -0,0 +1,1186 @@
+ * @return             True on success.
+ */
+bool BITCOINKERNEL_WARN_UNUSED_RESULT kernel_enable_log_category(const kernel_LogCategory category);
+
+/**
+ * Disable a specific log category for the global internal logger.
+ *
+ * @param[in] category If kernel_LOG_ALL is chosen, all categories will be disabled.
+ * @return             True on success.
+ */
+bool BITCOINKERNEL_WARN_UNUSED_RESULT kernel_disable_log_category(const kernel_LogCategory category);
+
+/**
+ * @brief Start logging messages through the provided callback. Log messages
+ * produced before this function is first called are buffered and on calling this
+ * function are logged immediately.
+ *
+ * @param[in] callback  Non-null, function through which messages will be logged.
+ * @param[in] user_data Nullable, holds a user-defined opaque structure. Is passed back
+ *                      to the user through the callback.
+ * @param[in] options   Sets formatting options of the log messages.
+ * @return              A new kernel logging connection, or null on error.
+ */
+kernel_LoggingConnection* BITCOINKERNEL_WARN_UNUSED_RESULT kernel_logging_connection_create(
+    kernel_LogCallback callback,
+    void* user_data,
+    const kernel_LoggingOptions options
+) BITCOINKERNEL_ARG_NONNULL(1);
+
+/**
+ * Stop logging and destroy the logging connection.
+ */
+void kernel_logging_connection_destroy(kernel_LoggingConnection* logging_connection);
+
+/**
+ * @brief Creates a chain parameters struct with default parameters based on the
+ * passed in chain type.
+ *
+ * @param[in] chain_type Controls the chain parameters type created.
+ * @return               An allocated chain parameters opaque struct.
+ */
+const kernel_ChainParameters* BITCOINKERNEL_WARN_UNUSED_RESULT kernel_chain_parameters_create(
+    const kernel_ChainType chain_type);
+
+/**
+ * Destroy the chain parameters.
+ */
+void kernel_chain_parameters_destroy(const kernel_ChainParameters* chain_parameters);
+
+/**
+ * @brief Creates an object for holding the kernel notification callbacks.
+ *
+ * @param[in] callbacks Holds the callbacks that will be invoked by the kernel notifications.
+ */
+kernel_Notifications* BITCOINKERNEL_WARN_UNUSED_RESULT kernel_notifications_create(
+    kernel_NotificationInterfaceCallbacks callbacks);
+
+/**
+ * Destroy the kernel notifications.
+ */
+void kernel_notifications_destroy(const kernel_Notifications* notifications);
+
+/**
+ * Creates an empty context options.
+ */
+kernel_ContextOptions* BITCOINKERNEL_WARN_UNUSED_RESULT kernel_context_options_create();
+
+/**
+ * @brief Sets the chain params for the context options. The context created
+ * with the options will be configured for these chain parameters.
+ *
+ * @param[in] context_options  Non-null, previously created with kernel_context_options_create.
+ * @param[in] chain_parameters Is set to the context options.
+ */
+void kernel_context_options_set_chainparams(
```

Comment:
There are a few places, like here, where we expose modifier functions that are (quasi) required to be ran before initializing another object. An alternative approach would be to extend the `kernel_context_options_create` to take a (nullable) `kernel_ChainParameters*`, and remove these ~unsafe modifiers altogether? I think that would have the benefit of:
- removing a whole category of bugs where users set options at the wrong time (i.e. too late), silently leading to buggy behaviour
- making it easier to see which options can (should) be set, without having to first have read the entire documentation

This concern also applies to e.g.:
- `kernel_context_options_set_notifications`
- `kernel_validation_interface_register`